Add CONTRIBUTING.md (#523)
chaoss/grimoirelab-sortinghat: CONTRIBUTING.md

# Contributing to GrimoireLab Sorting Hat

These are some general guidelines and information related to how we contribute to 
GrimoireLab. You can read about it from the [CONTRIBUTING.md](https://github.com/chaoss/grimoirelab/blob/master/CONTRIBUTING.md).

## Changelog Entries

Some of your contributions will require a changelog entry which explains 
the motivation of the change. These changelog entries should be written 

> to the `releases/unreleased/` directory. These entries would be included 

Reviewer: I'll move the sentence: `These changelog entries should be written to the `releases/unreleased/` directory.` as the first line after the section **Writing changelog entries". I'd add to that sentence a comment about that the entries are written in YAML format and should follow the next schema.

in the release notes to explain users and developers about the new features 
or bugs fixed in the software. This is an example of changelog entry:

```
title: 'Fix bug casting spells on magicians'
category: fixed
author: John Smith <jsmith@example.com>
issue: 666
notes: >
    The bug was making impossible to cast a spell on
    a magician.
```

Changelog entries will be written to explain *what* was changed and *why*, 
not *how*. Take into account not everybody is a developer and they are 
meant to reach a wider audience.

### What warrants a changelog entry

> The changelog entry is **required** for
> - Any change in the code that directly affects the GrimoireLab users
> - Bug fixes
> - New updates
> - Performance improvements

Reviewer: ```suggestion
Changelog entries are **required** for:
- Code changes that directly affects the GrimoireLab users.
- Bug fixes.
- New updates.
- Performance improvements.
```

> The changelog entry is **not required** for
> - Any docs-only (e.g., README.md) changes
> - Any developer-facing change (e.g., test suite changes)

Reviewer: ```suggestion
Changelog entries are **not required** for
- Docs-only (e.g., README.md) changes.
- Developer-facing change (e.g., test suite changes).
- Code refactoring.
```

### Writing changelog entries

```
title: 'Fix bug casting spells on magicians'
category: fixed
author: John Smith <jsmith@example.com>
issue: 666
notes: >
    The bug was making impossible to cast a spell on
    a magician.
```

The `issue` value is a reference to the issue, if any, that is targeted with 
this change, and the `author` key (format: `Your Name <YourName@example.org>`) 
is used to give attribution to community contributors. **Both are optional.** 
The `category` field maps the category of the change, valid options are: 
added, fixed, changed, deprecated, removed, security, performance, other. 
**Category field is mandatory.** The `notes` field should have a description 
explaining the changes in the code. Remember you can write blocks of text using 

> the `>` character at the beginning of each block. See the above given example.

Reviewer: I'd start explaining the title and the category fields. Then move to the other ones. A paragraph for each field.

> A good changelog entry should be descriptive and concise. It should explain the 

Reviewer: Can we find a better formatting for this part? It's not clear what each example is about.

Reviewer: It also needs to be a different section/sub-section.

Author: > Can we find a better formatting for this part? It's not clear what each example is about.

We can remove the bullet points and have plain paragraphs maybe. The other format could be a table but I think it doesn't look good to use a table here.

> It also needs to be a different section/sub-section.

Created a new sections `Tips` and `Examples`.

how it benefits the user. 

- **Bad:** Copy (some text) to clipboard.
- **Good:** Update the “Copy to clipboard” tooltip to indicate what’s being copied. 

Again, the first example is too vague and provides no context.

- **Bad:** Fixes and Improves CSS and HTML problems in mini pipeline graph and builds dropdown.
- **Good:** Fix tooltips and hover states in mini pipeline graph and builds dropdown. 

The first example is too focused on implementation details. The user doesn’t care 
that we changed CSS and HTML, they care about the result of those changes.

- **Bad:** Strip out `nil`s in the Array of Commit objects returned from `find_commits_by_message_with_elastic`
- **Good:** Fix 500 errors caused by Elasticsearch results referencing garbage-collected commits 

The first example focuses on *how* we fixed something, not on *what* it fixes. 
The rewritten version clearly describes the *end benefit* to the user 
(fewer 500 errors), and *when* (searching commits with Elasticsearch).

Use your best judgment and try to put yourself in the mindset of someone reading 
the compiled changelog. Does this entry add value? Does it offer context about 

> *what* and *why* the change was made?

Reviewer: Maybe this is more useful before the examples than after then.

Contributors can use the interactive [changelog](https://github.com/Bitergia/release-tools#changelog) 

> tool for this purpose which generates the changelog entry file automatically.

Reviewer: I think this part is more useful if it goes after the description of the fields.
